### DISCUSSION TITLE
Validation image size differs from training size  within each epoch

### DISCUSSION THREAD

**Discussion by SPONGEBOB-du** [ultralytics/ultralytics#20722] [Q&A] [answered]
Hi, I'm training a YOLOv8 model using the imgsz=320 argument. I expected both training and validation images to be resized to 320x320. However, during training, I noticed that the image sizes during validation differ from the training image size within each epoch.

Is there a way to force validation to always use the same size (320x320) as training?
Also, I am not using multi-scale training or custom dataloaders.

Here's the code I'm using:

`    results = model.train(data='coco128.yaml', 
                        epochs=2, 
                        batch=128,
                       device=[0,1],  # 指定GPU索引列表
                        workers=0,
                        imgsz=320,
                        )
`
![wechat_2025-05-20_131444_087](https://github.com/user-attachments/assets/b13d924a-d0ff-46b9-b240-af3925d0e207)
![wechat_2025-05-20_131426_551](https://github.com/user-attachments/assets/ba79f7e6-e9d9-488d-921b-9c41ff389605)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @SPONGEBOB-du, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response to help you get started quickly. An Ultralytics engineer will also assist you here soon 😊

**Reply by glenn-jocher** [reply]
Hi @SPONGEBOB-du,

This behavior occurs because by default, validation uses rectangular batching (`rect=True`) while training uses square images. The rectangular batching preserves aspect ratios during validation to improve efficiency and accuracy.

To force validation to use exactly 320×320 like your training, add the `rect=False` parameter:

```python
results = model.train(
    data='coco128.yaml', 
    epochs=2, 
    batch=128,
    device=[0,1],
    workers=0,
    imgsz=320,
    rect=False  # Force square images during validation
)
```

This will ensure both training and validation use the same 320×320 square image size throughout the process.

**Reply by SPONGEBOB-du** [reply]
I'm sorry, but I’ve already tried setting rect=False. However, during training, the validation image size still changes in each epoch. For example, I passed in images with shape [128, 3, 320, 320], but during validation, the input shape changes to [128, 3, 352, 352].

Is there a way to enforce a fixed validation image size throughout training?

**Comment by Y-T-G** [comment, marked as answer]
You can change `rect` to False here

https://github.com/ultralytics/ultralytics/blob/5e892b62255d9b26efe64a89389ac928ff7931e2/ultralytics/models/yolo/detect/train.py#L65

**Reply by SPONGEBOB-du** [reply]
thanks so much.  it works.

**Reply by glenn-jocher** [reply]
You're welcome! I'm glad that resolved the issue for you. The behavior you experienced is actually by design - during training, the validator defaults to `rect=True` for efficiency, but modifying that line in the source code gives you the exact control you needed over validation image sizes.
